File: gopher-mcp/sdk/typescript/mcp-example/src/mcp-calculator-client.ts
/**
 * @file mcp-calculator-client.ts
 * @brief Real working MCP Calculator Client with GopherTransport
 *
 * This is a complete working MCP client that connects to the calculator server
 * using GopherTransport with real C++ filter infrastructure.
 *
 */

import { ChainExecutionMode, RoutingStrategy } from "@mcp/filter-sdk";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { CallToolResultSchema, ListToolsResultSchema } from "@modelcontextprotocol/sdk/types.js";
import { GopherTransport, GopherTransportConfig } from "./gopher-transport";

async function main() {
  let transport: GopherTransport | null = null;

  try {
    console.log("🚀 Starting Calculator MCP Client with GopherTransport");

    // Create comprehensive GopherTransport configuration for client
    const transportConfig: GopherTransportConfig = {
      name: "calculator-client-transport",
      version: "1.0.0",
      protocol: "tcp", // Use TCP for real network communication
      host: "localhost", // Connect to localhost
      port: 8080, // Connect to server port 8080
      connectTimeout: 10000, // 10 second connection timeout
      sendTimeout: 5000, // 5 second send timeout
      receiveTimeout: 10000, // 10 second receive timeout

      // Client-specific filter configuration
      filters: {
        // Security filters for client
        security: {
          authentication: {
            method: "jwt",
            secret: "calculator-server-secret-key-2024", // Must match server secret
            issuer: "calculator-client",
            audience: "calculator-server",
          },
          authorization: {
            enabled: true,
            policy: "allow",
            rules: [
              {
                resource: "tools/*",
                action: "call",
                conditions: { authenticated: true },
              },
            ],
          },
        },

        // Observability for client
        observability: {
          accessLog: {
            enabled: true,
            format: "json",
            fields: ["timestamp", "method", "sessionId", "duration", "clientId", "operation"],
            output: "console",
          },
          metrics: {
            enabled: true,
            labels: {
              component: "mcp-client",
              transport: "gopher",
              service: "calculator",
              environment: "production",
            },
          },
          tracing: {
            enabled: true,
            serviceName: "calculator-client",
            samplingRate: 0.3, // 30% sampling for client
          },
        },

        // Traffic management for client
        trafficManagement: {
          rateLimit: {
            enabled: true,
            requestsPerMinute: 1000, // Lower rate limit for client
            burstSize: 50,
            keyExtractor: "custom",
          },
          circuitBreaker: {
            enabled: true,
            failureThreshold: 5, // Lower threshold for client
            timeout: 30000,
            resetTimeout: 60000,
          },
          retry: {
            enabled: true,
            maxAttempts: 3,
            backoffStrategy: "exponential",
            baseDelay: 1000,
            maxDelay: 5000,
          },
        },

        // Error handling
        errorHandling: {
          stopOnError: false,
          retryAttempts: 2,
          fallbackBehavior: "passthrough",
        },

        // Chain configuration for optimal performance
        chain: {
          executionMode: ChainExecutionMode.SEQUENTIAL, // Process filters in order
          routingStrategy: RoutingStrategy.ROUND_ROBIN,
          maxParallel: 1,
          bufferSize: 8192, // 8KB buffer
          timeoutMs: 15000, // 15 second timeout
        },

        // Buffer pool configuration (disabled for now)
        // bufferPool: {
        //   bufferSize: 4096,
        //   maxBuffers: 50,
        //   preallocCount: 5,
        //   useThreadLocal: true,
        //   zeroOnAlloc: false,
        // },

        // NEW: CApiFilter integration for real-time message processing
        customCallbacks: {
          onMessageReceived: message => {
            console.log(`🔍 [CApiFilter DEBUG] Client onMessageReceived called!`);
            console.log(
              `🔍 [CApiFilter DEBUG] Original message:`,
              JSON.stringify(message, null, 2)
            );
            console.log(`🔍 [CApiFilter DEBUG] Message type: ${message.method || "notification"}`);
            console.log(`🔍 [CApiFilter DEBUG] Message ID: ${message.id || "N/A"}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );

            // Add client-specific metadata
            if ("id" in message) {
              const processedMessage = {
                ...message,
                clientMetadata: {
                  receivedAt: Date.now(),
                  clientId: "calculator-client-001",
                  sessionId: "session-" + Math.random().toString(36).substr(2, 9),
                  processedBy: "CApiFilter-onMessageReceived",
                },
              };
              console.log(
                `🔍 [CApiFilter DEBUG] Processed message:`,
                JSON.stringify(processedMessage, null, 2)
              );
              console.log(`🔍 [CApiFilter DEBUG] Returning processed message to C++ filter chain`);
              return processedMessage;
            }
            return null; // No modification
          },
          onMessageSent: message => {
            console.log(`🔍 [CApiFilter DEBUG] Client onMessageSent called!`);
            console.log(
              `🔍 [CApiFilter DEBUG] Original message:`,
              JSON.stringify(message, null, 2)
            );
            console.log(`🔍 [CApiFilter DEBUG] Message type: ${message.method || "notification"}`);
            console.log(`🔍 [CApiFilter DEBUG] Message ID: ${message.id || "N/A"}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );

            // Add client-specific metadata
            if ("id" in message) {
              const processedMessage = {
                ...message,
                clientMetadata: {
                  sentAt: Date.now(),
                  clientId: "calculator-client-001",
                  sessionId: "session-" + Math.random().toString(36).substr(2, 9),
                  processedBy: "CApiFilter-onMessageSent",
                },
              };
              console.log(
                `🔍 [CApiFilter DEBUG] Processed message:`,
                JSON.stringify(processedMessage, null, 2)
              );
              console.log(`🔍 [CApiFilter DEBUG] Returning processed message to C++ filter chain`);
              return processedMessage;
            }
            return null; // No modification
          },
          onConnectionEstablished: connectionId => {
            console.log(`🔍 [CApiFilter DEBUG] Client onConnectionEstablished called!`);
            console.log(`🔍 [CApiFilter DEBUG] Connection ID: ${connectionId}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
          onConnectionClosed: connectionId => {
            console.log(`🔍 [CApiFilter DEBUG] Client onConnectionClosed called!`);
            console.log(`🔍 [CApiFilter DEBUG] Connection ID: ${connectionId}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
          onError: (error, context) => {
            console.log(`🔍 [CApiFilter DEBUG] Client onError called!`);
            console.log(`🔍 [CApiFilter DEBUG] Error:`, error.message);
            console.log(`🔍 [CApiFilter DEBUG] Context: ${context}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
        },
      },
    };

    transport = new GopherTransport(transportConfig);

    // Create a client
    const client = new Client({
      name: "calculator-client",
      version: "1.0.0",
    });

    // Start the GopherTransport (will connect to server)
    await transport.start();

    // Connect the client to the transport
    await client.connect(transport);

    console.log("✅ Connected to Calculator MCP Server via GopherTransport");
    console.log("🔐 Security: JWT authentication enabled");
    console.log("📊 Observability: Logging, metrics, and tracing enabled");
    console.log("⚡ Traffic Management: Rate limiting, circuit breaker, retry enabled");
    console.log(`📊 Transport stats:`, transport.getStats());

    // List available tools
    console.log("\n📋 Listing available tools...");
    try {
      const toolsResult = await client.request(
        { method: "tools/list", params: {} },
        ListToolsResultSchema
      );

      console.log("Available tools:");
      if (toolsResult.tools.length === 0) {
        console.log("  No tools available");
      } else {
        for (const tool of toolsResult.tools) {
          console.log(`  - ${tool.name}: ${tool.description}`);
        }
      }
    } catch (error) {
      console.log(`❌ Failed to list tools: ${error}`);
    }

    // Perform calculator operations
    console.log("\n🧮 Performing calculator operations...");

    const operations = [
      { operation: "add", a: 10, b: 5, description: "10 + 5" },
      { operation: "subtract", a: 20, b: 8, description: "20 - 8" },
      { operation: "multiply", a: 6, b: 7, description: "6 × 7" },
      { operation: "divide", a: 100, b: 4, description: "100 ÷ 4" },
      { operation: "power", a: 2, b: 8, description: "2^8" },
      { operation: "sqrt", a: 144, description: "√144" },
      { operation: "factorial", a: 5, description: "5!" },
    ];

    for (const op of operations) {
      try {
        console.log(`\n🔢 ${op.description}:`);

        const params: any = { operation: op.operation, a: op.a };
        if (op.b !== undefined) {
          params.b = op.b;
        }

        const result = await client.request(
          {
            method: "tools/call",
            params: {
              name: "calculator",
              arguments: params,
            },
          },
          CallToolResultSchema
        );

        if (result.content && result.content.length > 0) {
          console.log(`  Result: ${result.content[0].text}`);
        } else {
          console.log("  No result content");
        }
      } catch (error) {
        console.error(`  ❌ Error: ${error}`);
      }
    }

    // Test server statistics
    console.log("\n📊 Getting server statistics...");
    try {
      const statsResult = await client.request(
        {
          method: "tools/call",
          params: {
            name: "server_stats",
            arguments: { include_details: true },
          },
        },
        CallToolResultSchema
      );

      if (statsResult.content && statsResult.content.length > 0) {
        console.log("Server Statistics:");
        console.log(statsResult.content[0].text);
      }
    } catch (error) {
      console.error(`❌ Failed to get server stats: ${error}`);
    }

    // Test error handling
    console.log("\n⚠️  Testing error handling...");
    try {
      const errorResult = await client.request(
        {
          method: "tools/call",
          params: {
            name: "calculator",
            arguments: { operation: "divide", a: 10, b: 0 },
          },
        },
        CallToolResultSchema
      );

      if (errorResult.content && errorResult.content.length > 0) {
        console.log(`Error handling test: ${errorResult.content[0].text}`);
      }
    } catch (error) {
      console.error(`❌ Error handling test failed: ${error}`);
    }

    // Test rate limiting (send multiple requests quickly)
    console.log("\n⚡ Testing rate limiting...");
    const rateLimitPromises = [];
    for (let i = 0; i < 5; i++) {
      rateLimitPromises.push(
        client
          .request(
            {
              method: "tools/call",
              params: {
                name: "calculator",
                arguments: { operation: "add", a: i, b: 1 },
              },
            },
            CallToolResultSchema
          )
          .catch(error => ({ error: error.message }))
      );
    }

    const rateLimitResults = await Promise.all(rateLimitPromises);
    console.log(`Rate limit test: ${rateLimitResults.length} requests sent`);
    const successCount = rateLimitResults.filter(r => !r.error).length;
    const errorCount = rateLimitResults.filter(r => r.error).length;
    console.log(`  ✅ Successful: ${successCount}`);
    console.log(`  ❌ Failed: ${errorCount}`);

    console.log("\n✅ Calculator client operations completed successfully!");
    console.log(`📊 Final transport stats:`, transport.getStats());
  } catch (error) {
    console.error("❌ Client error:", error);
    process.exit(1);
  } finally {
    // Clean up
    try {
      if (transport) {
        await transport.close();
        console.log("🔌 Transport connection closed");
      }
    } catch (error) {
      console.error("❌ Error closing transport:", error);
    }
  }
}

// Start the client
main().catch(error => {
  console.error("❌ Failed to start client:", error);
  process.exit(1);
});

File: gopher-mcp/sdk/rust/mcp-example/src/mcp-calculator-client.rs
/**
 * @file mcp-calculator-client.rs
 * @brief Calculator Client with GopherTransport
 *
 * This is a calculator client that connects to the calculator server
 * using GopherTransport with real C++ filter infrastructure.
 *
 */

use mcp_filter_sdk::{
    GopherTransport, GopherTransportConfig, ProtocolType,
    CApiFilter, FilterCallbacks, FilterStatus, FilterManagerConfig,
    LibraryLoader
};
use serde_json::json;
use std::time::Duration;
use std::sync::Arc;
use tracing::{info, warn, error};

/// Calculator client implementation
pub struct CalculatorClient {
    transport: GopherTransport,
    client_filter: CApiFilter,
}

impl CalculatorClient {
    /// Create a new calculator client
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        // Initialize the library loader
        let library = Arc::new(LibraryLoader::new()?);
        
        // Create client-specific filter callbacks
        let callbacks = FilterCallbacks {
            on_data: Some(Box::new(|data, end_stream| {
                info!("📊 Client filter processing data: {} bytes, end_stream: {}", data.len(), end_stream);
                FilterStatus::Continue
            })),
            on_write: Some(Box::new(|data, end_stream| {
                info!("📤 Client filter writing data: {} bytes, end_stream: {}", data.len(), end_stream);
                FilterStatus::Continue
            })),
            on_new_connection: Some(Box::new(|state| {
                info!("🔗 Client filter: new connection with state: {}", state);
            })),
            on_error: Some(Box::new(|code, message| {
                error!("❌ Client filter error: {} - {}", code, message);
            })),
            on_high_watermark: Some(Box::new(|| {
                warn!("⚠️ Client filter: high watermark reached");
            })),
            on_low_watermark: Some(Box::new(|| {
                info!("📉 Client filter: low watermark reached");
            })),
            user_data: None,
        };
        
        // Create CApiFilter
        let client_filter = CApiFilter::new(
            library,
            callbacks,
            None,
        )?;
        
        let config = GopherTransportConfig {
            name: "calculator-client".to_string(),
            version: "1.0.0".to_string(),
            protocol: ProtocolType::Tcp,
            host: Some("localhost".to_string()),
            port: Some(8080),
            connect_timeout: Some(Duration::from_millis(30000)),
            send_timeout: Some(Duration::from_millis(5000)),
            receive_timeout: Some(Duration::from_millis(5000)),
            max_connections: Some(1),
            buffer_size: Some(8192),
            filter_config: Some(FilterManagerConfig {
                max_filters: 5,
                debug: true,
                metrics: true,
            }),
        };

        Ok(Self {
            transport: GopherTransport::new(config),
            client_filter,
        })
    }

    /// Connect to the calculator server
    pub async fn connect(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        info!("🚀 Starting Calculator MCP Client with GopherTransport");

        // Initialize the transport
        self.transport.start().await?;
        info!("✅ Transport started successfully");

        Ok(())
    }

    /// Run example calculations
    pub async fn run_examples(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        info!("🧮 Running calculator examples...");

        // Test different types of messages
        let messages = vec![
            json!({
                "jsonrpc": "2.0",
                "id": 1,
                "method": "add",
                "params": {
                    "a": 5,
                    "b": 3
                }
            }),
            json!({
                "jsonrpc": "2.0",
                "id": 2,
                "method": "subtract",
                "params": {
                    "a": 10,
                    "b": 4
                }
            }),
            json!({
                "jsonrpc": "2.0",
                "id": 3,
                "method": "multiply",
                "params": {
                    "a": 6,
                    "b": 7
                }
            }),
        ];

        for message in messages {
            self.transport.send(message).await?;
            info!("✅ Sent calculation request");
        }

        Ok(())
    }

    /// Disconnect from the server
    pub async fn disconnect(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        info!("🔌 Disconnecting from server...");
        
        self.transport.close().await?;
        info!("✅ Transport closed");
        
        info!("👋 Client disconnected successfully");
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_env_filter("info")
        .init();

    let mut client = CalculatorClient::new()?;
    
    // Connect to the server
    if let Err(e) = client.connect().await {
        error!("Failed to connect to server: {}", e);
        return Err(e);
    }

    // Run example calculations
    if let Err(e) = client.run_examples().await {
        error!("Failed to run examples: {}", e);
        client.disconnect().await?;
        return Err(e);
    }

    // Disconnect
    client.disconnect().await?;

    Ok(())
}